### DISCUSSION TITLE
Tests for connection pool?

### DISCUSSION THREAD

**Discussion by jzabroski** [dotnet/SqlClient#1686] [Q&A] [answered]
Hi,

I was looking for concurrency tests that validate the connection pooling implementation, but don't really see anything.  In fact, it looks like SqlConnection.ClearAllPools isn't even covered.

Can you point me to where the connection pooling implementation is code covered?

**Comment by JRahnama** [comment, marked as answer]
@jzabroski maybe looking into[ performance counter](https://docs.microsoft.com/en-us/sql/connect/ado-net/performance-counters?view=sql-server-ver16) will help you to test your implementation. There are some uncovered parts of the driver that we are working on, but with this driver it takes a lot to improve CodeCov even for 2%.

**Reply by jzabroski** [reply]
I appreciate the frank reply!

I kind of feel like I would just be more successful if I re-write the whole library, to be honest... similar to MimeKit killing off SmtpClient.

I was thinking using a protocol framework like BedrockFramework would make things arranged more easily as linear, independent combinations of "tabular data stream" ops. It would also allow injecting custom policies a lot easier, so that things like sp_reset_connection can be done at higher levels of abstraction if required, thus enabling even faster multi-threaded throughput as there can be fewer round trips per "session".

**Reply by ErikEJ** [reply]
@jzabroski a kind if rewrite/ rework is in progress.

**Reply by jzabroski** [reply]
For what it's worth, You can learn a lot about db driver design by talking to the guy at Progress Software who wrote the JDBC drivers. He had a talk I sat through when in college in 2008 where he articulated, quite thoughtfully, the evolution of ODBC/JDBC driver design by breaking it into "generations".  This happened around the same time as UT-Austin computer scientist William Cook presenting his research papers on "remote batch execution" for data store access, and Michael Stonebreaker advocating Vertica DB use a Ruby-like DSL for columnar-store querying - and such ideas actually came to fruition via kdb+ vendor libraries like [Exxeleron's qSharp](https://github.com/exxeleron/qSharp) C# library.

Anyway, a better library actually has real cost savings and advantages for these large-scale data science solutions that would significantly boost gross margin when you consider things like AWS API Rate Limits on EC2, IAM, S3, etc.

Java and Java Vendors are remarkably more open on design of their drivers, c.f. https://docs.progress.com/bundle/datadirect-connect-jdbc-51/page/Welcome-to-the-Progress-DataDirect-Connect-Series-for-JDBC.html
